### PR TITLE
Implement ambiguous and imaginary datetime handling (PEP 495)

### LINKED ISSUE #72: Wrong result when a shift involves a DST change.
In paris in 2013, we changed to DST in 2013-03-31 (at 2 AM it was 3 AM), from UTC+1 to UTC+2 (DST)

Here's an example of the issue I have with arrow : 

``` python
>>> import arrow, datetime

>>> just_before = arrow.get(datetime.datetime(2013, 03, 31, 1, 50, 45), "Europe/Paris").ceil("hour")
>>> just_before
<Arrow [2013-03-31T01:59:59.999999+01:00]>

>>> just_after = just_before.replace(microseconds=+1)
>>> just_after
<Arrow [2013-03-31T02:00:00+02:00]>
>>> # That is not right... It should be either 2 AM UTC+1 or preferably  3AM UTC +2
>>> # One could argue that depending on what you ask for, it may be the correct answer, but then the following is :

>>> (just_after - just_before).total_seconds()
1e-06
>>> # That is right but not consistent with the dates

>>> (just_after.to("utc") - just_before.to("utc").total_seconds()
-3599.999999
# That should be the same value as the previous computation. Plus, the value is negative because :

>>> just_before.to('utc'), just_after.to('utc')
(<Arrow [2013-03-31T00:59:59.999999+00:00]>,
 <Arrow [2013-03-31T00:00:00+00:00]>)
```

I think the problem is pretty clear...

### LINKED ISSUE #315: Daylight savings bug
I have an embedded system which has no  Real Time Clock, running Debian Jessie.  On boot the time defaults to 1-Jan-1970 UTC.  I am using `chrony` to syncrhonise the system time with NTP servers.

My data logging app is running as a daemon (via `systemd`) and waits for the time to sync before logging (checks system time is > 2015).  Both UTC and Local Time are logged and the Local Time is reporting **AEST (Australian Eastern STANDARD Time)** instead of **AEDT (Australian Eastern DAYLIGHT-SAVINGS Time)**.

If I type `date` it reports the correct local time (AEDT).  If if run a python shell and run `arrow.now()` it reports the correct local time (AEDT).  If I restart my app (after network time is synced) then the app logs the correct local time (AEDT).

**Is this a bug in arrow (v0.4.4) ?**

One possible work around for me is to not start the app until network time is synchronised, but I'm not confident that arrow will report the correct local time after daylight savings ends.

Thanks for any help to solve this issue.
Brendan.

### LINKED ISSUE #368: Daylight savings time bug
I was trying out some stuff with timezones, and I encountered the following bug:

```
In [58]: a = arrow.get('2016-10-30 00:00')

In [59]: a
Out[59]: <Arrow [2016-10-30T00:00:00+00:00]>

In [60]: b = a.to('Europe/Amsterdam')

In [61]: b
Out[61]: <Arrow [2016-10-30T02:00:00+01:00]>

In [62]: c = b.to('UTC')

In [63]: a.to('UTC')
Out[63]: <Arrow [2016-10-30T00:00:00+00:00]>

In [64]: b
Out[64]: <Arrow [2016-10-30T02:00:00+01:00]>

In [65]: c
Out[65]: <Arrow [2016-10-30T01:00:00+00:00]>
```

In other words, I think 'b' should still have UTC offset +02:00 instead of +01:00, because the daylight savings time switches at 03:00 local time. So the offset can only be +01:00 starting from 01:00 UTC time.

### LINKED ISSUE #424: No way to specify whether it's DST.
For datetimes that are ambiguous - that is, ones that are in the hour before/after switching **from** DST - there's no way to specify which side they fall on. Neither `arrow.get` nor `arrow.arrow.Arrow` supports this.

For example:
```python
# Are these 02:00 before or after turning the clock back? Before, according to Arrow. Always.
Arrow(2017, 10, 29, 2, 0, tzinfo='Europe/Stockholm')
arrow.get(datetime(2017, 10, 29, 2, 0), 'Europe/Stockholm')
```

Pytz supports specifying this in the `tzinfo.localize` function via the `is_dst` parameter - for example, `timezone("Europe/Stockholm").localize(datetime(2017, 10, 29, 2, 0), is_dst=False)`. I propose Arrow solve it the same way by adding an `is_dst` keyword argument to both `arrow.arrow.Arrow` and `arrow.get`. Maybe someone's got a better idea, but those are my 2¢.

### LINKED ISSUE #476: fall dst transition bug?
I'm trying to solve a crash that occurred in my code last night and I'm hoping that switching to the arrow library may fix my problem. I've written a test script to confirm arrow's behavior during the dst transition and I think there may be a bug in my logic or in arrow.

The attached test script is generating the following output. Notice that when the transition happens, I get a time delta that looks like about 47 hours:

d=2017-11-05T01:59:54.990000-05:00, d2=2017-11-05T01:59:55.985000-05:00, delta=0:00:00.995000
d=2017-11-05T01:59:55.985000-05:00, d2=2017-11-05T01:59:56.993000-05:00, delta=0:00:01.008000
d=2017-11-05T01:59:56.993000-05:00, d2=2017-11-05T01:59:57.996000-05:00, delta=0:00:01.003000
d=2017-11-05T01:59:57.996000-05:00, d2=2017-11-05T01:59:59-05:00, delta=0:00:01.004000
d=2017-11-05T01:59:59-05:00, d2=2017-11-05T01:00:00.011000-05:00, delta=**-1 day, 23:00:01.011000**
d=2017-11-05T01:00:00.011000-05:00, d2=2017-11-05T01:00:01.012000-05:00, delta=0:00:01.001000
d=2017-11-05T01:00:01.012000-05:00, d2=2017-11-05T01:00:02.014000-05:00, delta=0:00:01.002000
d=2017-11-05T01:00:02.014000-05:00, d2=2017-11-05T01:00:03.019000-05:00, delta=0:00:01.005000
d=2017-11-05T01:00:03.019000-05:00, d2=2017-11-05T01:00:04.034000-05:00, delta=0:00:01.015000

[dst_test.zip](https://github.com/crsmithdev/arrow/files/1444239/dst_test.zip)

### LINKED ISSUE #509: Broken behaviour across DST transitions / Two things that look the same aren't
Possibly similar problem to #476, but this is about date time calculations rather than coping when the system clock changes.

While writing a snippet on calculating [time differences in python](https://bitbucket.org/snippets/maxgrenderjones/9e7nM6), I came across the following bizarre (to me) behaviour that looks like it might be a bug. 

The relevant lines from the snippet are shown below:

```
>>> from arrow.arrow import Arrow
>>> before = Arrow(2017, 10, 29, tzinfo='Europe/London')
>>> before + timedelta(days=1)
<Arrow [2017-10-30T00:00:00+00:00]>
>>> Arrow(2017, 10, 30, tzinfo='Europe/London')
<Arrow [2017-10-30T00:00:00+00:00]>
```

Spot the difference? I can't. Indeed:

```
>>> before + timedelta(days=1) - Arrow(2017, 10, 30, tzinfo='Europe/London')
datetime.timedelta(0)
```

So they're the same, right? Let's substitute them in then:

```
>>> (before + timedelta(days=1) - before).total_seconds() / 3600
24.0
>>> (Arrow(2017, 10, 30, tzinfo='Europe/London') - before).total_seconds() / 3600
25.0
```

(Note that 25 is the correct answer).

### LINKED ISSUE #686: Problem with differences between time points
Hi,

the ``shift()`` method is nice and seems to work with localized time with some kind of "natural" semantic. However it leads to some results (related to differences between time points).
 ```
>>> t0 = arrow.now().shift(weeks=3)    # The day before the next DST fall transition

>>> t0
<Arrow [2019-10-26T16:33:24.883042+02:00]>

>>> t1 = t0.shift(days=1)

>>> t1
<Arrow [2019-10-27T16:33:24.883042+01:00]>

>>> t1 - t0
datetime.timedelta(1)
```
So far, so good. A DST transition occurs between ``t0`` and ``t1`` (I'm living in Paris) and shifting by one day keeps the same time (16:33) with UTC offset adjusted (+02:00 to +01:00). The difference is one day, expressed as a ``timedelta``.

Now, convert the dates into UTC: 

```
>>> t2 = t0.to('UTC')

>>> t2
<Arrow [2019-10-26T14:33:24.883042+00:00]>

>>> t3 = t1.to('UTC')

>>> t3
<Arrow [2019-10-27T15:33:24.883042+00:00]>

>>> t2 == t0 and t3 == t1
True

>>> t1 - t2
datetime.timedelta(1, 3600)

>>> (t3 - t2) == (t1 - t0)
False

>>> t0 + (t3 - t2) == t1
False
```
The problem is that, while the time points are the same in local time or UTC (which is expected), their difference is no longer the same. And a few expected invariants are no longer true.

As a conclusion, 
1. Differences between time points shall always imply a conversion to UTC, **even if they are in the same time zone**. At least, we would have ``t1 - t0 == t3 - t2 == t3 - t0 == t1 - t2``. This should be clearly explained in the documentation.
2. The documentation should explain that a difference between 2 time points lead to a "naive" delta, even if ``arrow`` enforces time zone aware time points.
3. Time zone aware time delta could be introduced with a clean semantic (to be defined?)...

Best regards,

Antoine

### LINKED ISSUE #690: DST transition again
`
>>> t = arrow.get("2019-10-27T02:21:01+03:00")
>>> t
<Arrow [2019-10-27T02:21:01+03:00]>
>>> t.to("Israel")
<Arrow [2019-10-27T01:21:01+03:00]>
`
Note the same tz offset but different hours

### PR DESCRIPTION
**Pull request by systemcatch**
## Pull Request Checklist

- [x] 🧪 Added **tests** for changed code.
- [x] 🛠️ All tests **pass** when run locally (run `tox` or `make test` to find out!).
- [ ] 📚 Updated **documentation** for changed code.
- [x] ⏩ Code is **up-to-date** with the `master` branch.

## Description of Changes
This PR adds support for handling ambiguous and imaginary datetimes via the fold attribute [PEP 495](https://www.python.org/dev/peps/pep-0495/) and dateutil [methods](https://dateutil.readthedocs.io/en/stable/tz.html#dateutil.tz.enfold). I'm happy to admit that without the `dateutil.tz.enfold()` method this feature would have been much harder to backport into 2.7/3.5.

There are several design decisions I've made here that need to be looked at carefully.

- There is now a requirement for python-dateutil >= 2.7.0, before it was simply required.
- The `arrow.Arrow()` constructor will issue a warning but not prevent creation of imaginary datetimes.
- Various other methods will prevent creation of these imaginary datetimes and will resolve them to existing datetimes with no warning to the user.

### TODO
- [ ] Decide what to do with `span` method, `interval`, `span_range`, `ceil` and `floor` all depend upon it.
- [ ] Understand `Pacific/Apia` edge case.
- [ ] Consider adding a method to automatically resolve imaginary datetimes.
- [ ] Cleanup code.
- [ ] Bulk out tests and link to issues.
- [x] Add pendulum dst example, coverage 100%.
- [ ] Documentation updates where needed.

ref #488 
closes #72 
closes #315 
closes #368 
closes #424 
closes #476 
closes #509 
closes #686 
closes #690

### CODE REVIEW

File: arrow/arrow.py [hunk 1]
```diff
@@ -79,10 +90,22 @@ def __init__(
...
+            fold=self._fold,
         )
 
+        # warn user if dt is imaginary
+        if not dateutil_tz.datetime_exists(self._datetime):
+            warnings.warn(
+                "{} does not exist in the current timezone - {}".format(
+                    self._datetime, tzinfo
+                ),
+                util.ImaginaryDatetimeWarning,
```
Comment (krisfremen) [inline arrow/arrow.py:106]: would be nice to have this exposed as a property or variable instead of just a warning.

Comment (systemcatch) [reply arrow/arrow.py:106]: Good idea, I think using a property is the way to go.

**Comment by systemcatch** [comment]
Closing this in favour of a new (and hopefully correct! 😆 ) PR I'm working on.
